### ISSUE TITLE
package names versus folder names

### ISSUE THREAD

**Issue by mark-hahn** [atom/atom#5608]
I get quite confused when packages use names in their repo like `atom-my-package` when the Atom package name is actually `my-package`.  I end up renaming folders several times to clone a repo, work on it, and do a PR.

Enough bitching and on to the problem.  When I had  the `atom-my-package` folder in `~/.atom/packages` then the package listed in settings was`atom-my-package`.  But the settings were saved under  `my-package` which is what was in `package.json`.  So settings didn't work until I renamed the folder `my-package`.

All code in atom should use the package name in `package.json` or they should all use the folder name containing package.json.  Right now they are not consistent.

**Comment by izuzak** [comment]
> When I had the atom-my-package folder in ~/.atom/packages then the package listed in settings wasatom-my-package. But the settings were saved under my-package which is what was in package.json. So settings didn't work until I renamed the folder my-package.

@mark-hahn Can you provide exact steps which reproduce this behavior? I have a package with the name "pdf-view" which is stored in a repository called "atom-pdf-view" (different name), and I don't see this problem: Settings View show "pdf-view" as the name and the `.atom/packages` folder has `pdf-view` as a subfolder:

https://github.com/izuzak/atom-pdf-view

https://atom.io/packages/pdf-view

![screen shot 2015-02-18 at 11 57 26](https://cloud.githubusercontent.com/assets/38924/6246083/5740f79a-b765-11e4-9b0d-ce5d80036b6c.png)

```
$ ls .atom/packages/
total 0
drwxr-xr-x   8 izuzak   272B Feb 16 10:26 atom-space-race/
drwxr-xr-x   6 izuzak   204B Feb 16 10:26 atom-space-race-syntax/
drwxr-xr-x  10 izuzak   340B Feb 18 11:59 pdf-view/
```

**Comment by thomasjo** [comment]
Similar to @izuzak, I have never experienced this either.

As an example my LaTeX package is named `latex`, but the repository is named `atom-latex` (both on GitHub and locally). APM correctly names the soft-link `latex` for dev mode, and also gives it the correct folder name — `latex` — when installed "proper".

**Comment by postcasio** [comment]
I think the issue is with the name on disk, not specifically the repository name. I've seen this happen with development packages mainly. I have a package named `printer` that I develop locally in a directory named `atom-printer`. When it's linked into `~/.atom/dev/packages`, its name is `atom-printer` instead of `printer`.

**Comment by izuzak** [comment]
@postcasio Can you turn that description into a list of steps which reproduce the problem for you, just to make sure I'm following this? I'm seeing this:

```
$ cat package.json | grep name
  "name": "pdf-view",

$ apm link --dev
/Users/izuzak/.atom/dev/packages/pdf-view -> /Users/izuzak/github/atom-pdf-view
```

If I understood correctly, you're saying that you observed this:

```
$ apm link --dev
/Users/izuzak/.atom/dev/packages/atom-pdf-view -> /Users/izuzak/github/atom-pdf-view
```

**Comment by postcasio** [comment]
Ahh, I hadn't actually linked the package. If I do `apm link` it correctly creates the symlink using the name in `package.json`. But I often just clone repositories directly into `~/.atom/dev/packages`. So the steps are:

1. `cd ~/.atom/dev/packages`
2. `git clone https://github.com/postcasio/atom-printer`
3. `atom -d atom-printer`

The package name in the settings view is `atom-printer`, and config values are saved under `atom-printer`, but as far as the package is concerned, its name is `printer` and it will look under `printer` for config values etc.

So basically, this only happens if you're manually messing around with your installed packages - if you have a package in a directory named differently to its intended name, Atom will take the directory name as canonical instead of the name in `package.json`.

**Comment by mark-hahn** [comment]
> I think the issue is with the name on disk

Definitely.

>  Atom will take the directory name as canonical instead of the name in
package.json.

That is the "bug" that caused me to create this issue.  Some of Atom uses
the folder name and part uses package.json.

So there is an undocumented rule that package.json should match the folder
name.  The rule should be "fixed" or Atom should show an error and not load
the package.

**Comment by izuzak** [comment]
> So there is an undocumented rule that package.json should match the folder name.  The rule should be "fixed" or Atom should show an error and not load the package.

@mark-hahn As @postcasio said -- you shouldn't really be poking inside Atom's configuration folders and expect it to behave in a certain way. Instead, you should use `apm link` and `apm unlink` to link/unlink your development folder into Atom. Is there a reason why you can't use the workflow for which there are specific apm commands?

1. create a development folder outside `~/.atom`
2. use `apm link` and `apm unlink` to link that folder into Atom

I'm not sure if I would call this a bug and I'm not convinced we should do anything about this.  Instead, I'd rather like to improve the documentation so that users don't feel encouraged to perform such rocket-surgery and guide them towards supported workflows. @postcasio @mark-hahn Does that sound good to you?

**Comment by postcasio** [comment]
Yes, I agree. Would be nice to fix but I don't think it's a pressing issue.

**Comment by mark-hahn** [comment]
> Is there a reason why you can't use the workflow for which there are
specific apm commands?

Can APM clone a repo and start using it?  What if the repo name doesn't
match the package name?

I'll do some more testing and tell exactly what behavior seems like a bug
to me.

**Comment by postcasio** [comment]
> Can APM clone a repo and start using it?  What if the repo name doesn't
match the package name?

You can use `apm develop package-name` to clone an Atom package and automatically link it into `~/.atom/dev/packages`. It works regardless of the repository name.

**Comment by izuzak** [comment]
> Can APM clone a repo and start using it?

Yep:

```
$ apm dev --help 
Usage: apm develop <package_name> [<directory>]

Clone the given package's Git repository to the directory specified,
install its dependencies, and link it for development to
~/.atom/dev/packages/<package_name>.

If no directory is specified then the repository is cloned to
~/github/<package_name>. The default folder to clone packages into can
be overridden using the ATOM_REPOS_HOME environment variable.

Once this command completes you can open a dev window from atom using
cmd-shift-o to run the package out of the newly cloned repository.

Options:
  --help, -h  Print this usage message
```

> What if the repo name doesn't match the package name?

See my comment above in which a package I have has a different repo name (atom-pdf-view) from the name in package.json (pdf-view). Everything works. https://github.com/atom/atom/issues/5608#issuecomment-74893315

**Comment by mark-hahn** [comment]
I was referring to a community package, not a core package.  Here is the sequence to reproduce one of several bugs.

- cd to ~/.atom/packages
- git clone https://github.com/izuzak/atom-pdf-view.git
- cd atom-pdf-view
- npm install
- atom --dev .
- Open settings
- click on packages
- search for pdf
  - wrong package name of `atom-pdf-view` appears
- click on `atom-pdf-view`
- `atom-pdf-view` settings appear, again with wrong name
- click on `atom-pdf-view` at top
.

Nothing happens and the console shows this error ..

![jjhg](https://cloud.githubusercontent.com/assets/811455/6305767/49e0eaaa-b8e2-11e4-9d72-d35fe92748ee.gif)

**Comment by kevinsawicki** [comment]
> npm install

You want `apm install`

**Comment by kevinsawicki** [comment]
That is why the incompatible native module icon is shown in the status bar

**Comment by izuzak** [comment]
> I was referring to a community package, not a core package.

I was referring to a community package as well.

> cd to ~/.atom/packages
> ...
> wrong package name of atom-pdf-view appears

@mark-hahn Can you explain the _need_ to manually create/change things in Atom's ~/.atom/packages folder? Why is cloning to a folder **outside** the ~/.atom folder and using `apm link` not a workflow you can use?

1. cd ~
2. git clone https://github.com/izuzak/atom-pdf-view.git
3. cd atom-pdf-view
4. apm link or apm link --dev
5. apm install
6. start Atom and continue with other steps

**Comment by mark-hahn** [comment]
I originally tried apm install and got the error below so I switch to npm install.

```
C:\Users\Administrator\.atom\packages\atom-pdf-view>apm install
Installing modules failed
npm ERR! Failed resolving git HEAD (https://github.com/mozilla/pdf.js.git) fatal: ambiguous argument 'v1.0.907': unknown revi
sion or path not in the working tree.
npm ERR! Failed resolving git HEAD (https://github.com/mozilla/pdf.js.git) Use '--' to separate paths from revisions, like th
is:
npm ERR! Failed resolving git HEAD (https://github.com/mozilla/pdf.js.git) 'git <command> [<revision>...] -- [<file>...]'
npm ERR! Failed resolving git HEAD (https://github.com/mozilla/pdf.js.git)
npm ERR! Windows_NT 6.2.9200
npm ERR! argv "C:\\Users\\Administrator\\AppData\\Local\\atom\\app-0.181.0\\resources\\app\\apm\\bin\\\\node.exe" "C:\\Users\
\Administrator\\AppData\\Local\\atom\\app-0.181.0\\resources\\app\\apm\\node_modules\\npm\\bin\\npm-cli.js" "--globalconfig"
"C:\\Users\\Administrator\\.atom\\.apm\\.apmrc" "--userconfig" "C:\\Users\\Administrator\\.atom\\.apmrc" "install" "--target=
0.21.0" "--arch=ia32" "--msvs_version=2013"
npm ERR! node v0.10.35
npm ERR! npm  v2.5.1
npm ERR! code 128

npm ERR! Command failed: fatal: ambiguous argument 'v1.0.907': unknown revision or path not in the working tree.
npm ERR! Use '--' to separate paths from revisions, like this:
npm ERR! 'git <command> [<revision>...] -- [<file>...]'
npm ERR!
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <http://github.com/npm/npm/issues>
```

**Comment by izuzak** [comment]
> I originally tried apm install and got the error below so I switched to npm install.

Can you try a different package? That's exception is different issue, related to https://github.com/atom/apm/pull/292

**Comment by mark-hahn** [comment]
> Why is cloning to a folder outside the ~/.atom folder and using apm link not a workflow you can use?

I've never used an external folder because `apm link` didn't work in windows when I first used Atom.  I'll reverse the question.  Why is using the projects folder a workflow I can't use?  I've used it for 9 months.

**Comment by izuzak** [comment]
> Can you try a different package? That's exception is different issue, related to atom/apm#292

Ahh it's not related to that, but it's definitely a different issue.

**Comment by mark-hahn** [comment]
> I was referring to a community package as well.

You use the dev folder for community packages?  I thought that was for core packages.  

So we have three locations to consider.  The dev folder, the packages folder, and an external folder.  What are the rules about these folders?

**Comment by izuzak** [comment]
> Why is using the projects folder a workflow I can't use? 

Because you're having expectation outside of officially supported behavior. The ~/.atom folder is folder for Atom, a program, not you, a human. If you want to use those folders as a human - you should follow the same rules followed by Atom when it uses that folder. If you don't follow the rules, then you're on your own.

**Comment by izuzak** [comment]
> The dev folder, the packages folder, and an external folder. What are the rules about these folders?

The recommendation is: use `apm link` and `apm link --dev` to link folders from outside of `~/.atom` into it.

**Comment by mark-hahn** [comment]
I repeated the test with `atom-the-closer`.  Showing the steps after Atom opened ...

- click on packages
- search for `closer`
  - wrong package name of `atom-the-closer` appears
- click on `atom-the-closer`
- `atom-the-closer` settings appear, again with wrong name
- click on `atom-pdf-view` at top

This time there was no error in the console but nothing happened.

> The recommendation is: use apm link and apm link --dev to link folders from outside of ~/.atom into it.

OK, let's assume I'm breaking some rule.  What is that rule?  In other words what should I change in the sequence above?

> The recommendation is: use apm link and apm link --dev to link folders from outside of ~/.atom into it.

OK, I'll try that as a test.

**Comment by izuzak** [comment]
> What is that rule? In other words what should I change in the sequence above?

Do what `apm link` does -- it uses the name of the package from package.json as the name of the folder in `~/.atom/packages` or `~/.atom/dev/packages`.

**Comment by mark-hahn** [comment]
That worked.  So now I know the rule.  User package folders must be outside of `~/.atom`.  I will remember that.  

I'm sorry for the trouble but you can see how it looked broken.  Especially after using Atom for so long with everything in the packages folder.

**Comment by mark-hahn** [comment]
BTW, it still bugs me that the name of the folder doesn't match the package name.  It makes looking for them harder.  Looking for `atom-xxx` after not finding it where it belongs isn't very hard.  But the developer could name the folder something totally unrelated.  Oh well.

**Comment by izuzak** [comment]
> User package folders must be outside of ~/.atom. I will remember that.

I wouldn't say it's a must. The recommendation is more general: use supported tools and workflows instead of manually creating/changing things within Atom's ~/.atom folder. The bump in the road you hit might be a bug and Atom really should use the name from package.json if it's available. But there might be other bumps. That's why I'm encouraging you to use supported tools and workflows -- you avoid those bumps altogether. 

> I'm sorry for the trouble but you can see how it looked broken. 

No trouble at all. It's helpful to see what causes problems, friction or confusion. And I do understand why this was confusing. I'll also try to figure out if the current behavior around this is intentional and if it can be changed so that other users don't get confused, and also improve documentation. :bow:
